core/src/main/java/org/apache/calcite/rel/metadata/FunctionalDependencySet.java

```diff
@@ -0,0 +1,369 @@
+  }
+
+  /**
+   * Compute the minimal cover of this functional dependency set.
+   * Returns an equivalent set with minimal dependencies.
+   */
+  public FunctionalDependencySet minimalCover() {
+    // Split multi-attribute right sides into single attributes
+    Set<FunctionalDependency> splitFDs = new HashSet<>();
+    for (FunctionalDependency fd : fdSet) {
+      splitFDs.addAll(fd.split());
+    }
+    splitFDs.removeIf(FunctionalDependency::isTrivial);
+
+    // Remove redundant attributes from left sides
+    Set<FunctionalDependency> reducedFDs = new HashSet<>();
+    for (FunctionalDependency fd : splitFDs) {
+      FunctionalDependencySet tempSet = new FunctionalDependencySet(splitFDs);
+      tempSet.removeFD(fd);
+      reducedFDs.add(reduceLeft(fd, tempSet));
+    }
+
+    // Remove redundant functional dependencies
+    reducedFDs.removeIf(fd -> {
+      FunctionalDependencySet remainingFDs = new FunctionalDependencySet(reducedFDs);
+      remainingFDs.removeFD(fd);
+      return remainingFDs.implies(fd.getDeterminants(), fd.getDependents());
+    });
+
+    return new FunctionalDependencySet(reducedFDs);
+  }
+
+  /**
+   * Reduce left side by removing redundant columns from determinants.
+   */
+  private static FunctionalDependency reduceLeft(FunctionalDependency fd,
+      FunctionalDependencySet fdSet) {
+    ImmutableBitSet determinants = fd.getDeterminants();
+    ImmutableBitSet dependents = fd.getDependents();
+
+    // Try removing each attribute to find minimal determinant set
+    for (int attr : fd.getDeterminants()) {
+      ImmutableBitSet reduced = determinants.clear(attr);
+      if (fdSet.closure(reduced).contains(dependents)) {
+        determinants = reduced;
+      }
+    }
+    return new FunctionalDependency(determinants, dependents);
+  }
+
+  /**
+   * Check if this FD set is equivalent to another FD set.
+   * Two FD sets are equivalent if they have the same closure for any attribute set.
+   */
+  public boolean equalTo(FunctionalDependencySet other) {
+    // Check if every FD in this set is implied by the other set
+    for (FunctionalDependency fd : fdSet) {
+      if (!other.implies(fd.getDeterminants(), fd.getDependents())) {
+        return false;
+      }
+    }
+    // Check if every FD in the other set is implied by this set
+    for (FunctionalDependency fd : other.fdSet) {
+      if (!implies(fd.getDeterminants(), fd.getDependents())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Find all candidate keys within the given attribute set.
+   * A candidate key is a minimal subset of the given attributes such that
+   * its closure contains all the given attributes.
+   *
+   * @param attributes the set of attributes to search for candidate keys within
+   * @return a set of minimal attribute subsets that can determine all given attributes
+   */
+  public Set<ImmutableBitSet> findCandidateKeys(ImmutableBitSet attributes) {
```

Comment:
The number of keys can be exponential in the size of the attributes in the worst case.
I don't know how often this happens in practice.

Reply:
This is indeed a problem. I can only say that when I tested these test cases on my local Mac, they were all in milliseconds, less than 1 second. I haven't thought of a better test scenario yet.